File: tanstia/tanstia/page/mail_campaigns/mail_campaigns.py
import frappe
from frappe import _

@frappe.whitelist()
def get_data(type_new=None, region=None, district=None, association_category=None, association_name=None):
    filters = {'docstatus':['!=',2]}
    if type_new:
        filters["type"] = type_new
    if region:
        filters["region"] = region
    if district:
        filters["district"] = district
    if association_category:
        filters["association_category"] = association_category
    if association_name:
        filters["name"] = association_name
    records = frappe.db.get_all(
        "Association",
        filters=filters,
        fields=["name","type","association_category","region","status","member_id",'status','member_name','type','date_and_year_of_affiliation']
    )

    if not records:
        return "<p class='text-muted'>No records found</p>"

    html = """
    <style>
        .scrollable-table-container {
            max-height: 600px;
            overflow-y: auto;
            border: 1px solid #ccc;
        }
        table {
            width: 100%;
            border-collapse: collapse !important;
        }
        table, th, td {
            border: 1px solid black !important;
        }
        thead th {
            background-color: #0F1568 !important;
            color: white !important;
            text-align: center;
            font-size: 14px;
            position: sticky;
            top: 0;
            z-index: 2;
        }
        
    </style>

    <div class="scrollable-table-container">
        <table class="table-hover mb-0" id="members-table">
            <thead>
                <tr>
                    <th>
                        <input type="checkbox" id="select-all">
                    </th>
                    <th>S.No</th>
                    <th>Member ID</th>
                    <th>Member Name</th>
                    <th>Association</th>
                    <th>Position</th>
                    <th>Status</th>
                    <th>Type</th>
                    <th>Region</th>
                    <th>Office Bearer</th>
                    <th>Elected <br>EC Members</th>
                    <th>Email</th>
                    
                </tr>
            </thead>
            <tbody>
    """

    sno = 1
    for parent in records:
        child_doc = frappe.db.get_all(
            'Office Bearer Details',
            {
                'parent': parent.name,
                'parentfield': 'office_bearers',
                'parenttype': 'Association'
            },
            ['office_bearer', 'designation', 'email', 'contact_number']
        )

        for row in child_doc:
            html += f"""
                <tr>
                    <td style="text-align:center;">
                        <input type="checkbox" class="row-checkbox"
                               data-email="{row.email or ''}">
                    </td>
                    <td style="text-align:center;">{sno}</td>
                    <td>{parent.member_id or ''}</td>
                    <td>{parent.member_name or ''}</td>
                    <td> <a href="/app/association/{parent.name or '' }">
                        { parent.name or '' }
                    </a></td>
                    
                    <td>{row.designation or ''}</td>
                    <td>{parent.status or ''}</td>
                    <td>{parent.type or ''}</td>
                    <td>{parent.region or ''}</td>
                    <td>{row.office_bearer or ''}</td>
                    
                    <td>{parent.elected_ec_members_size or ''}</td>
                    <td>{row.email or ''}</td>
                </tr>
            """
            sno += 1
    
        office_bearers = frappe.db.get_all(
            'Office Bearer Details',
            {
                'parentfield': 'office_bearers',
                'parenttype': 'EC Office Bearer'
            },
            ['office_bearer', 'designation', 'email', 'contact_number']
        )

        for row in office_bearers:
            html += f"""
                <tr>
                    <td style="text-align:center;">
                        <input type="checkbox" class="row-checkbox"
                               data-email="{row.email or ''}">
                    </td>
                    <td style="text-align:center;">{sno}</td>
                    <td></td>
                    <td>{parent.member_name or ''}</td>
                    <td></td>
                    
                    <td>{row.designation or ''}</td>
                    <td></td>
                    <td></td>
                    <td>{row.office_bearer or ''}</td>
                    <td></td>
                    <td></td>
                    <td>{row.email or ''}</td>

                </tr>
            """
            sno += 1
    
        records = frappe.db.get_all(
        "Association",
        filters=filters,
        fields=["name"]
    )
    html += """
            </tbody>
        </table>
    </div>

    <script>
        (function () {
            const selectAll = document.getElementById("select-all");
            const rowCheckboxes = document.querySelectorAll(".row-checkbox");

            selectAll.addEventListener("change", function () {
                rowCheckboxes.forEach(cb => cb.checked = selectAll.checked);
            });

            rowCheckboxes.forEach(cb => {
                cb.addEventListener("change", function () {
                    selectAll.checked =
                        document.querySelectorAll(".row-checkbox:checked").length === rowCheckboxes.length;
                });
            });
        })();
    </script>
    """

    return html

# @frappe.whitelist()
# def send_mail_campaign(subject, content, recipients, attachments=None):

#     if isinstance(recipients, str):
#         recipients = frappe.parse_json(recipients)

#     if isinstance(attachments, str):
#         attachments = frappe.parse_json(attachments)

#     # ✅ Build proper attachment list
#     email_attachments = []

#     if attachments:
#         for file_name in attachments:
#             file_doc = frappe.get_doc("File", file_name)

#             email_attachments.append({
#                 "fname": file_doc.file_name,
#                 "fcontent": file_doc.get_content()
#             })

#     MAX_SIZE = 500

#     unique_recipients = list(
#         {email.strip().lower() for email in recipients if email}
#     )

#     campaign = frappe.new_doc("Campaign")
#     campaign.created_by = frappe.session.user
#     campaign.triggered_email_count = len(unique_recipients)
#     campaign.in_progress_count = len(unique_recipients)

#     for i in range(0, len(unique_recipients), MAX_SIZE):
#         batch = unique_recipients[i:i + MAX_SIZE]
#         frappe.errprint(email_attachments)
#         email_queue = frappe.sendmail(
#             recipients=batch,
#             subject=subject,
#             message=content,
#             attachments=email_attachments,  
#             delayed=True
#         )

#         for email in batch:
#             campaign.append("recepients", {
#                 "email": email,
#                 "status": "Sending",
#                 "email_queue": email_queue.name
#             })

#     campaign.insert(ignore_permissions=True)
#     return campaign.name



@frappe.whitelist()
def send_mail_campaign(subject, content, recipients, attachments=None):

    # ----------------------------
    # Normalize inputs
    # ----------------------------
    if isinstance(recipients, str):
        recipients = frappe.parse_json(recipients)

    if isinstance(attachments, str):
        attachments = frappe.parse_json(attachments)

    recipients = list({e.strip().lower() for e in recipients if e})

    if not recipients:
        frappe.throw("No valid recipients found")

    # ----------------------------
    # Prepare attachments
    # ----------------------------
    mail_attachments = []

    if attachments:
        for file_name in attachments:
            file_doc = frappe.get_doc("File", file_name)

            mail_attachments.append({
                "fname": file_doc.file_name,
                "fcontent": file_doc.get_content()
            })

    # ----------------------------
    # Send in batches
    # ----------------------------
    MAX_SIZE = 500
    frappe.errprint(attachments)
    for i in range(0, len(recipients), MAX_SIZE):
        batch = recipients[i:i + MAX_SIZE]

        frappe.sendmail(
            recipients=batch,
            subject=subject,
            message=content,
            attachments=mail_attachments,
            delayed=False   
        )

    return "Email sent successfully"

File: tanstia/tanstia/page/mail_campaigns/mail_campaigns.js
// Global variables
let type_new, region, district, association_category, association_name;

frappe.pages['mail-campaigns'].on_page_load = function(wrapper) {
    const page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Mail Campaigns',
        single_column: true
    });

    let $headerContainer = $('<div class="form-group d-flex justify-content-between align-items-end mb-3" style="gap: 20px;">')
        .appendTo(page.main);

    let $filtersSection = $('<div class="d-flex" style="gap: 15px; flex-wrap: wrap;">')
        .appendTo($headerContainer);

    let $buttonsSection = $('<div class="d-flex" style="gap: 10px;">')
        .appendTo($headerContainer);

    // Type
    type_new = frappe.ui.form.make_control({
        df: {
            fieldtype: 'Select',
            label: 'Type',
            fieldname: 'type',
            options: ['', 'Member', 'Non-Member']
        },
        parent: $filtersSection,
        render_input: true
    });

    // Region
    region = frappe.ui.form.make_control({
        df: {
            fieldtype: 'Link',
            label: 'Region',
            options: 'Region',
            fieldname: 'region'
        },
        parent: $filtersSection,
        render_input: true
    });

    // District
    district = frappe.ui.form.make_control({
        df: {
            fieldtype: 'Link',
            label: 'District',
            options: 'District',
            fieldname: 'district'
        },
        parent: $filtersSection,
        render_input: true
    });

    // Association Category
    association_category = frappe.ui.form.make_control({
        df: {
            fieldtype: 'Link',
            label: 'Association Category',
            options: 'Association Category',
            fieldname: 'association_category'
        },
        parent: $filtersSection,
        render_input: true
    });

    // Association
    association_name = frappe.ui.form.make_control({
        df: {
            fieldtype: 'Link',
            label: 'Association',
            options: 'Association',
            fieldname: 'association_name'
        },
        parent: $filtersSection,
        render_input: true
    });

    // Send Button
    let send = frappe.ui.form.make_control({
        df: {
            fieldtype: 'Button',
            label: 'Send Email',
            fieldname: 'send'
        },
        parent: $filtersSection,
        render_input: true
    });
    $(send.wrapper).css('margin-top', '25px');
    // Queries
    district.df.get_query = () => {
        if (region.get_value()) {
            return { filters: { region: region.get_value() } };
        }
    };

    association_name.df.get_query = () => {
        let filters = {};
        if (region.get_value()) filters.region = region.get_value();
        if (association_category.get_value()) filters.association_category = association_category.get_value();
        return { filters };
    };

    region.$input.on("change", () => {
        district.set_value("");
    });
    let $btnWrapper = $('<div style="margin-top:25px; display:flex; gap:10px;">')
    .appendTo($buttonsSection);

    let applyBtn = $(
        `<button class="btn btn-primary" style="margin-top:25px;">
            <i class="fa fa-filter"></i> Filter
        </button>`
    ).appendTo($btnWrapper);

    let clearBtn = $(
        `<button class="btn btn-secondary" style="margin-top:25px;">
            &#x2715;
        </button>`
    ).appendTo($btnWrapper);


    // Buttons
    // let applyBtn = $(`<button class="btn btn-primary"><i class="fa fa-filter"></i> Filter</button>`)
    //     .appendTo($buttonsSection);
    //     $(applyBtn.wrapper).css('margin-top', '20px');

    // let clearBtn = $(`<button class="btn btn-secondary">&#x2715;</button>`)
    //     .appendTo($buttonsSection);
    $(clearBtn.wrapper).css('margin-top', '25px');
    let $container = $('<div class="mt-4">').appendTo(page.main);

    // Apply
    applyBtn.on("click", function () {
        get_data();
    });

    // Clear
    clearBtn.on("click", function () {
        type_new.set_value("");
        region.set_value("");
        district.set_value("");
        association_category.set_value("");
        association_name.set_value("");
        $container.empty();
        frappe.call({
            method: "tanstia.tanstia.page.mail_campaigns.mail_campaigns.get_data",
            args: {
            },
            callback: function (r) {
                if (r.message) {
                    $(".mt-4").html(r.message);
                }
            }
        });
    });

    // Send Email
    send.$input.on("click", function () {
        open_send_email_dialog();
    });

    // Initial load
    get_data();
};

// Fetch data
function get_data() {
    frappe.call({
        method: "tanstia.tanstia.page.mail_campaigns.mail_campaigns.get_data",
        args: {
            type_new: type_new.get_value(),
            region: region.get_value(),
            district: district.get_value(),
            association_category: association_category.get_value(),
            association_name: association_name.get_value()
        },
        callback: function (r) {
            if (r.message) {
                $(".mt-4").html(r.message);
            }
        }
    });
}


// // Send email dialog
// function open_send_email_dialog() {

//     /* ---------------------------
//        Collect emails
//     ---------------------------- */
//     let selectedEmails = [];

//     document.querySelectorAll(".row-checkbox:checked").forEach(cb => {
//         if (cb.dataset.email) selectedEmails.push(cb.dataset.email);
//     });

//     if (!selectedEmails.length) {
//         document.querySelectorAll(".row-checkbox").forEach(cb => {
//             if (cb.dataset.email) selectedEmails.push(cb.dataset.email);
//         });
//     }

//     /* ---------------------------
//        "Child table" storage
//     ---------------------------- */
//     let attachment_rows = [];

//     function render_attachments(dialog) {
//         let html = attachment_rows.map((f, i) => `
//             <div style="margin-bottom:6px">
//                 📎 <a href="${f.file_url}" target="_blank">
//                     ${f.file_name}
//                 </a>
//                 <span style="cursor:pointer;color:red;margin-left:8px"
//                     onclick="remove_page_attachment(${i})">✖</span>
//             </div>
//         `).join("");

//         dialog.fields_dict.attached_files.$wrapper.html(
//             html || "<i>No attachments</i>"
//         );
//     }

//     window.remove_page_attachment = function (idx) {
//         attachment_rows.splice(idx, 1);
//         render_attachments(dialog);
//     };

//     let dialog = new frappe.ui.Dialog({
//         title: 'Send Email',
//         size: 'large',
//         fields: [
//             { fieldtype: 'Data', fieldname: 'subject', label: 'Subject', reqd: 1 },
//             { fieldtype: 'Text Editor', fieldname: 'content', label: 'Content', reqd: 1 },

//             // 👇 Attach used only as trigger
//             { fieldtype: 'Attach', fieldname: 'attachment', label: 'Attach File' },

//             // 👇 Multi-file preview
//             { fieldtype: 'HTML', fieldname: 'attached_files' }
//         ],

//         primary_action_label: 'Send',
//         primary_action(values) {
//             console.log(attachment_rows)
//             frappe.call({
//                 method: "tanstia.tanstia.page.mail_campaigns.mail_campaigns.send_mail_campaign",
//                 args: {
//                     subject: values.subject,
//                     content: values.content,
//                     recipients: selectedEmails,
//                     attachments: attachment_rows.map(f => f.name)
//                 },
//                 callback() {
//                     frappe.msgprint("Email sent successfully");
//                     dialog.hide();
//                 }
//             });
//         }

//     });

//     dialog.show();

//     /* ---------------------------
//        Attach onchange handler
//     ---------------------------- */
//     dialog.fields_dict.attachment.df.onchange = () => {

//         let file_url = dialog.get_value('attachment');
//         if (!file_url) return;

//         // Get File doc using file_url
//         frappe.call({
//             method: "frappe.client.get_list",
//             args: {
//                 doctype: "File",
//                 filters: { file_url },
//                 fields: ["name", "file_name", "file_url"],
//                 limit_page_length: 1
//             },
//             callback(r) {
//                 if (!r.message || !r.message.length) return;

//                 let file = r.message[0];

//                 // ✅ Append to "child table"
//                 attachment_rows.push({
//                     name: file.name,          // File docname
//                     file_name: file.file_name,
//                     file_url: file.file_url   // Correct private/public URL
//                 });

//                 // ✅ Clear Attach field
//                 dialog.set_value('attachment', null);

//                 // ✅ Refresh UI
//                 render_attachments(dialog);
//             }
//         });
//     };
// }


function open_send_email_dialog() {

    /* ---------------------------
       Collect emails
    ---------------------------- */
    let selectedEmails = [];

    document.querySelectorAll(".row-checkbox:checked").forEach(cb => {
        if (cb.dataset.email) selectedEmails.push(cb.dataset.email);
    });

    if (!selectedEmails.length) {
        document.querySelectorAll(".row-checkbox").forEach(cb => {
            if (cb.dataset.email) selectedEmails.push(cb.dataset.email);
        });
    }

    /* ---------------------------
       Attachment storage
    ---------------------------- */
    let attachment_rows = [];

    function render_attachments(dialog) {
        let html = attachment_rows.map((f, i) => `
            <div style="margin-bottom:6px">
                📎 <a href="${f.file_url}" target="_blank">${f.file_name}</a>
                <span style="cursor:pointer;color:red;margin-left:8px"
                    data-idx="${i}" class="remove-attachment">✖</span>
            </div>
        `).join("");

        dialog.fields_dict.attached_files.$wrapper.html(
            html || "<i>No attachments</i>"
        );
    }

    /* ---------------------------
       Dialog
    ---------------------------- */
    let dialog = new frappe.ui.Dialog({
        title: 'Send Email',
        size: 'large',
        fields: [
            { fieldtype: 'Data', fieldname: 'subject', label: 'Subject', reqd: 1 },
            { fieldtype: 'Text Editor', fieldname: 'content', label: 'Content', reqd: 1 },

            { fieldtype: 'Button', fieldname: 'upload_files', label: 'Add Attachments' },
            { fieldtype: 'HTML', fieldname: 'attached_files' }
        ],

        primary_action_label: 'Send',
        primary_action(values) {
            frappe.call({
                method: "tanstia.tanstia.page.mail_campaigns.mail_campaigns.send_mail_campaign",
                args: {
                    subject: values.subject,
                    content: values.content,
                    recipients: selectedEmails,
                    attachments: attachment_rows.map(f => f.name)
                },
                callback() {
                    frappe.msgprint("Email sent successfully");
                    dialog.hide();
                }
            });
        }
    });

    dialog.show();

    /* ---------------------------
       Remove attachment
    ---------------------------- */
    dialog.fields_dict.attached_files.$wrapper.on(
        'click',
        '.remove-attachment',
        function () {
            attachment_rows.splice($(this).data('idx'), 1);
            render_attachments(dialog);
        }
    );

    /* ---------------------------
       Multi-file uploader
    ---------------------------- */
    dialog.fields_dict.upload_files.$input.on('click', () => {
        new frappe.ui.FileUploader({
            allow_multiple: true,
            on_success(file) {
                attachment_rows.push({
                    name: file.name,
                    file_name: file.file_name,
                    file_url: file.file_url
                });
                render_attachments(dialog);
            }
        });
    });
}


